<?xml version="1.0"?>
<!--
Software License Agreement (BSD)

\file      client_msg_shims.launch
\authors   Dave Niewinski <dniewinski@clearpathrobotics.com>
\copyright Copyright (c) 2017, Clearpath Robotics, Inc., All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that
the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the
   following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the 
   following disclaimer in the documentation and/or other materials provided with the distribution.
 * Neither the name of Clearpath Robotics nor the names of its contributors may be used to endorse or promote
   products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WAR-
RANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, IN-
DIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->
<launch>
  <arg name="platform_control_odom" default="/platform_control/odom" />
  <arg name="platform_odom" default="/platform_odometry" />
  <arg name="front_laser" default="/front/scan" />
  <arg name="rear_laser" default="/rear/scan" />
  <arg name="stereo_camera" default="/stereo_camera" />
  <arg name="cmd_vel" default="/platform_control/cmd_vel" />
  <arg name="diagnostics" default="/diagnostics" />

  <!-- Changing the line below may affect ARK performance. DO NOT CHANGE -->
  <node name="clock_throttler" type="throttle" pkg="topic_tools" args="messages /ark_bridge/clock_un_throttled 100 /ark_bridge/clock" />

  <node name="tf_bridge" pkg="ark_bridge" type="tf_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/tf"/>
    <param name="ros_topic" value="/tf"/>
    <param name="direction" value="ros2lcm"/>
  </node>
  <node name="tf_static_bridge" pkg="ark_bridge" type="tf_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/tf_static"/>
    <param name="ros_topic" value="/tf_static"/>
    <param name="direction" value="ros2lcm"/>
  </node>
  <node name="odom_bridge_1" pkg="ark_bridge" type="odom_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/platform_control_odom"/>
    <param name="ros_topic" value="$(arg platform_control_odom)"/>
    <param name="direction" value="ros2lcm"/>
  </node>
  <node name="odom_bridge_2" pkg="ark_bridge" type="odom_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/platform_odometry"/>
    <param name="ros_topic" value="$(arg platform_odom)"/>
    <param name="direction" value="ros2lcm"/>
  </node>
  <node name="twist_bridge" pkg="ark_bridge" type="twist_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/platform_control_cmd_vel"/>
    <param name="ros_topic" value="$(arg cmd_vel)"/>
    <param name="direction" value="lcm2ros"/>
  </node>
  <node name="clock_bridge" pkg="ark_bridge" type="clock_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/clock_un_throttled"/>
    <param name="ros_topic" value="/clock"/>
    <param name="direction" value="ros2lcm"/>
  </node>
  <node name="slam_map_bridge" pkg="ark_bridge" type="occupancygrid_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/slam_map"/>
    <param name="ros_topic" value="/map"/>
    <param name="direction" value="lcm2ros"/>
  </node>

  <node name="front_scan_bridge" pkg="ark_bridge" type="laserscan_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/front_scan"/>
    <param name="ros_topic" value="$(arg front_laser)"/>
    <param name="direction" value="ros2lcm"/>
  </node>
  <node name="rear_scan_bridge" pkg="ark_bridge" type="laserscan_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/rear_scan"/>
    <param name="ros_topic" value="$(arg rear_laser)"/>
    <param name="direction" value="ros2lcm"/>
  </node>

  <node name="stereo_bridge" pkg="ark_bridge" type="pointcloud2_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/stereo_camera_0"/>
    <param name="ros_topic" value="$(arg stereo_camera)"/>
    <param name="direction" value="ros2lcm"/>
  </node>

  <node name="diagnostics_bridge" pkg="ark_bridge" type="diagnosticarray_bridge" output="screen" respawn="true">
    <param name="lcm_topic" value="/ark_bridge/ark_diagnostics"/>
    <param name="ros_topic" value="$(arg diagnostics)"/>
    <param name="direction" value="lcm2ros"/>
  </node>
</launch>
